mersenne-31/src/dft.rs

```diff
@@ -0,0 +1,256 @@
+//! Implementation of DFT for `Mersenne31`.
+//!
+//! Strategy follows: https://www.robinscheibler.org/2013/02/13/real-fft.html
+//! In short, fold a Mersenne31 DFT of length n into a Mersenne31Complex DFT
+//! of length n/2. Some pre/post-processing is necessary so that the result
+//! of the transform behaves as expected wrt the convolution theorem etc.
+
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+use p3_dft::TwoAdicSubgroupDft;
+use p3_field::{AbstractField, Field, TwoAdicField};
+use p3_matrix::dense::RowMajorMatrix;
+use p3_matrix::{Matrix, MatrixRowSlices, MatrixRows};
+use p3_util::log2_strict_usize;
+
+use crate::{Mersenne31, Mersenne31Complex};
+
+/// Given an hxw matrix M = (m_{ij}) where h is even, return an
+/// (h/2)xw matrix N whose (k,l) entry is
+///
+///    Mersenne31Complex(m_{2k,l}, m_{2k+1,l})
+///
+/// i.e. the even rows become the real parts and the odd rows become
+/// the imaginary parts.
+///
+/// This packing is suitable as input to a Fourier Transform over the
+/// domain Mersenne31Complex; it is inverse to `idft_postprocess()`
+/// below.
+fn dft_preprocess(
+    input: RowMajorMatrix<Mersenne31>,
+) -> RowMajorMatrix<Mersenne31Complex<Mersenne31>> {
+    assert!(input.height() % 2 == 0, "input height must be even");
+    RowMajorMatrix::new(
+        input
+            .rows()
+            .tuples()
+            .flat_map(|(row_0, row_1)| {
+                // For each pair of rows in input, convert each
+                // two-element column into a Mersenne31Complex
+                // treating the first row as the real part and the
+                // second row as the imaginary part.
+                row_0
+                    .iter()
+                    .zip(row_1)
+                    .map(|(&x, &y)| Mersenne31Complex::new(x, y))
+            })
+            .collect(),
+        input.width(),
+    )
+}
+
+/// Transform the result of applying the DFT to the packed
+/// `Mersenne31` values so that the convolution theorem holds.
+///
+/// Source: https://www.robinscheibler.org/2013/02/13/real-fft.html
```

Comment:
It seems like this differs from the reference in that this doesn't bother appending the "redundant" `N/2 - 1` elements? Maybe worth noting